json-database/groups.json

```diff
@@ -0,0 +1,754 @@
+            "name": "The UK Symfony Meetup",
+            "meetup_com_slug": "symfony",
+            "country": "United Kingdom"
+        },
+        {
+            "name": "GlasgowPHP",
+            "meetup_com_slug": "GlasgowPHP",
+            "country": "United Kingdom"
+        },
+        {
+            "name": "PHP Roundtable Budapest",
+            "meetup_com_slug": "PHP-Roundtable-Budapest",
+            "country": "Hungary"
+        },
+        {
+            "name": "Veszprém PHP Meetup",
+            "meetup_com_slug": "Veszprem-PHP-Meetup",
+            "country": "Hungary"
+        },
+        {
+            "name": "PHP North East",
+            "meetup_com_slug": "PHP-North-East",
+            "country": "United Kingdom"
+        },
+        {
+            "name": "Louisville Laravel Meetup",
+            "meetup_com_slug": "Louisville-Laravel-Meetup",
+            "country": "Colorado"
+        },
+        {
+            "name": "Tampa Bay PHP",
+            "meetup_com_slug": "tampa-bay-php",
+            "country": "Florida"
+        },
+        {
+            "name": "Tampa Bay Laravel",
+            "meetup_com_slug": "Laravel",
+            "country": "Florida"
+        },
+        {
+            "name": "Boston PHP",
+            "meetup_com_slug": "bostonphp",
+            "country": "Massachusetts"
+        },
+        {
+            "name": "Detroit Laravel Meetup",
+            "meetup_com_slug": "Laravel-Detroit",
+            "country": "Michigan"
+        },
+        {
+            "name": "Kansas City PHP User Group",
+            "meetup_com_slug": "kcphpug",
+            "country": "Missouri"
+        },
+        {
+            "name": "North Jersey PHP & Programmers\" Convergence",
+            "meetup_com_slug": "nnjppc",
+            "country": "New Jersey"
+        },
+        {
+            "name": "Nashville PHP",
+            "meetup_com_slug": "nashvillephp",
+            "country": "Tennessee"
+        },
+        {
+            "name": "azPHP",
+            "meetup_com_slug": "azPHPUG",
+            "country": "Arizona"
+        },
+        {
+            "name": "Péhápkaři Brno",
+            "meetup_com_slug": "Pehapkari-Brno",
+            "country": "Czech Republic"
+        },
+        {
+            "name": "ColumbusPHP",
+            "meetup_com_slug": "phpphp",
+            "country": "Ohio"
+        },
+        {
+            "name": "AtlantaPHP",
+            "meetup_com_slug": "atlantaphp",
+            "country": "Georgia"
+        },
+        {
+            "name": "Symfony User Group Osnabrück",
+            "meetup_com_slug": "sfugos",
+            "country": "Germany"
+        },
+        {
+            "name": "BeerPHP Moscow",
+            "meetup_com_slug": "BeerPHP-Moscow",
+            "country": "Russia"
+        },
+        {
+            "name": "Meetup Laravel Lyon",
+            "meetup_com_slug": "Meetup-Laravel-Lyon",
+            "country": "France"
+        }
+    ]
+]
```

Comment:
@honzajavorek Btw, this is 150 items in json database :) 

The DB layer is https://github.com/donjajo/php-jsondb, that works like mysql (select, from, where...). Node/Python will have probably something similar and better.

The insert speed is 10 000 items in ~300 ms (https://donjajo.com/how-i-append-data-into-json-at-constant-time-php).

Reply:
Looks cool! Continuing the in-person discussion we had, I'll try to describe what I have now and why:

Scrapers using https://scrapy.org/ download data and spit them out as JSONL files. The data is not post-processed or cleaned, it's a raw output from the scrapers.

Those files are permanently stored on disk and accumulated using CircleCI cache with a constant caching key (retention to be implemented later, they have 500 MB limit on cache :D )

In another script, I read the files and from individual items I create db entries. This is heavily parallelized using https://docs.python.org/3/library/multiprocessing.html goodies like Pool or Queue. After I parse the item, but before I turn it into a db entry, I can do all sorts of post-processing (e.g. clean some values, re-structure some properties, parse text into words for further processing, geolocate where the item is from, and so on). This allows me to separate scraping from post-processing. That means I can keep the raw data and easily change the post-processing pipeline and keep running it on top of the very same data and it produces different things (easier development, but also migrations if something changes in the future, like schema).

One big advantage of putting the items into db is that the rest of the website works on top of that db (so that's uniform interface for like 20 other models with items from various other sources), but also, it gives me the query language, and, _tadada 🎉_, the indexes. I scrape raw items and save them per day, but there's a lot of duplicity. I want to keep the duplicity in the raw phase (track record of what the scraper actually downloaded on which day), but I want to merge the items for further work as part of the post-processing. The merging itself is based on a unique db index.

As an example, imagine I scrape job postings. The scraper downloads 50 job postings every day, but most of them will be the same as the previous day. Their URL will be the same. However, this isn't all! One job board can use various utm parameters. One company can advertise the same job on multiple job boards. This means I need to first post-process the data and reduce them into a synthetic identifier, then merge them according to that identifier. The architecture I strive for allows me to change the algorithm for synthetizing the identifier later, if I realize the one I have is not sufficient. At the same time, the actual mechanics of merging the items are dead simple, I just synthetize the ID, save the item, and if db complains on the unique index, I can pick up the existing entry and intelligently merge the items (e.g. keep properties from the newer one, or merge lists of URLs together).

Perhaps convoluted, but it's an iteration on top of what I already have, i.e. there are some lessons learned I have and which I already bake in to the architecture. I wish it was simpler, but at the same time, what I'm building is a product nobody else has :D

What next? Well I'm almost done with the post-processing and saving phase. Then I'll get back to the rest of the codebase I bulldozered in the past weeks, and I'll try to fix it according to the changes made, and make it work again. I might find difficulties in situations I forgot I should take care of.

Then, with my friend @miiila we want to adapt the system so it's able to use ML to distinguish junior job postings from non-junior ones. This is our long-term goal, which wasn't possible due to the limits of the previous architecture. With the new architecture, I hope it's gonna be finally possible to do it.

Then, I still might find technical limits due to the amount of data. I already design the scripts so that they can run as parallel jobs in CircleCI, which wasn't the case previously. That can buy me some time. If raw data and actual processing will be separate, I can also run the scrapers just nightly.

Another thing is storage. The JSONL files can become large. I might need to start using gzip for storing the HTML of the job descriptions, which takes up most of the space (but is necessary for evaluating juniority of the job postings). One of the beautiful characteristics of the new design is that if I feel the need, I can easily run separate scripts, which make one-time idempotent changes to all the raw data (adds gzip to certain properties, migrates schema), even the past data. I can also put the data to S3 or GitHub Pages, and then stream the files over network during the build. That would give me total manual control over the raw data files. The new architecture shouldn't prevent such change.

Well, enough writing, getting back to work 😅

Reply:
The PR in progress is here, but you probably won't see much there, it's a ball of mess https://github.com/honzajavorek/junior.guru/pull/780

Reply:
Thanks for writing up :+1: and taking a break :laughing: 

So it's 3 separates proceses?

* scaping jobs (medium)
* post processing them to filter out junior ones (hard)
* generate static webpage (easy)

That already emerges pretty clear architecture. Good luck with powering up :+1: 

---

Btw, the Github Pages allow up to 1 GB (https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits). For ~30 meetups json items it's ~30 kb (https://github.com/TomasVotruba/friendsofphp.org/blob/main/json-database/meetups.json), so there should be enough space for couple thousands items.

Reply:
How would be life easier if there would be single public json with all items :D 

I'm trying to walk the talk by sharing all the data used on the website:

* https://friendsofphp.org/api/meetups.json
* https://friendsofphp.org/api/groups.json

Reply:
Yup, the three phases are as you described it (simplified, e.g. there's additional process which takes the data from db and posts new jobs to Discord).

I use GitHub Pages for storing podcast files now, so I'm already educated about how much it can hold 😂

Tunneling the static data out as JSON API for anyone to use, that is awesome.